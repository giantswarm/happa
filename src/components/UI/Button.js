import PropTypes from 'prop-types';
import React from 'react';
import BsButton from 'react-bootstrap/lib/Button';

import LoadingIndicator from './LoadingIndicator';

// Button
//
// <Button
//   type='button|submit|reset'
//   bsStyle='primary|success|danger|info|warning|link'
//   loading=true|false
//   disabled=true|false
//   onClick=function>
//
//   Button Text
// </Button>
//
// A basic button. Can go into a 'loading' state, which will disable the button
// and show a spinner next to it.
//
// You can also disable the button by setting the disabled prop to true.
//

const Button = props => {
  const {
    loadingPosition,
    loading,
    disabled,
    bsStyle,
    bsSize,
    onClick,
    type,
    children,
    ...rest
  } = props;

  return (
    <div className='progress_button--container'>
      {loadingPosition === 'left' ? (
        <LoadingIndicator loading={loading} loadingPosition={loadingPosition} />
      ) : (
        undefined
      )}

      <BsButton
        bsSize={bsSize}
        bsStyle={bsStyle}
        disabled={disabled || loading}
        onClick={onClick}
        type={type}
        {...rest}
      >
        {children}
      </BsButton>

      {loadingPosition === 'right' ? (
        <LoadingIndicator loading={loading} loadingPosition={loadingPosition} />
      ) : (
        undefined
      )}
    </div>
  );
};

Button.propTypes = {
  type: PropTypes.string,
  bsStyle: PropTypes.string,
  bsSize: PropTypes.string,
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  loadingPosition: PropTypes.string,
  children: PropTypes.node,
};

Button.defaultProps = {
  loadingPosition: 'right',
};

export default Button;
