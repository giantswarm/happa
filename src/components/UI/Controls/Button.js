import PropTypes from 'prop-types';
import React from 'react';
import BsButton from 'react-bootstrap/lib/Button';
import LoadingIndicator from 'UI/Display/Loading/LoadingIndicator';

// Button
//
// <Button
//   type='button|submit|reset'
//   bsStyle='primary|success|danger|info|warning|link'
//   loading=true|false
//   disabled=true|false
//   onClick=function>
//
//   Button Text
// </Button>
//
// A basic button. Can go into a 'loading' state, which will disable the button
// and show a spinner next to it.
//
// You can also disable the button by setting the disabled prop to true.
//

const Button = (props) => {
  const {
    loadingPosition,
    loading,
    disabled,
    bsStyle,
    bsSize,
    onClick,
    type,
    children,
    loadingTimeout,
    ...rest
  } = props;

  return (
    <div className='progress_button--container'>
      {loadingPosition === 'left' ? (
        <LoadingIndicator
          loading={loading}
          loadingPosition={loadingPosition}
          timeout={loadingTimeout}
        />
      ) : undefined}

      <BsButton
        bsSize={bsSize}
        bsStyle={bsStyle}
        disabled={disabled || loading}
        onClick={onClick}
        type={type}
        role='button'
        {...rest}
      >
        {children}
      </BsButton>

      {loadingPosition === 'right' ? (
        <LoadingIndicator
          loading={loading}
          loadingPosition={loadingPosition}
          timeout={loadingTimeout}
        />
      ) : undefined}
    </div>
  );
};

Button.propTypes = {
  type: PropTypes.string,
  bsStyle: PropTypes.string,
  bsSize: PropTypes.string,
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  loadingPosition: PropTypes.string,
  children: PropTypes.node,
  className: PropTypes.string,
  loadingTimeout: PropTypes.number,
};

Button.defaultProps = {
  loadingPosition: 'right',
};

export default Button;
