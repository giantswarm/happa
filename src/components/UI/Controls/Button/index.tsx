import PropTypes from 'prop-types';
import React from 'react';
import BsButton from 'react-bootstrap/lib/Button';
import styled from 'styled-components';
import LoadingIndicator from 'UI/Display/Loading/LoadingIndicator';

const Wrapper = styled.div`
  .btn {
    border: 0px;
    border-radius: 3px;
    color: #fff;
    padding: 6px 12px;
    font-weight: 400;
    outline: none;
    cursor: pointer;
    transition: background-color 0.1s cubic-bezier(0, -0.41, 0.19, 1.44);
  }
  .btn.small {
    font-size: 14px;
    padding: 8px 18px;
  }

  .btn-secondary {
    background-color: #52728a;
    border: 1px #52728a solid;
  }

  .btn-primary {
    background-color: #2d9a2c;
    border: 1px #2d9a2c solid;
  }

  .btn-link {
    color: #ccc;
  }

  .btn-link:hover {
    color: #fff;
  }

  .btn:hover,
  .btn:active,
  .btn:active:focus,
  .btn:focus {
    background-color: #5c7f9a;
    color: #fff;
  }

  .btn-primary.disabled.focus,
  .btn-primary.disabled:focus,
  .btn-primary.disabled:hover,
  .btn-primary[disabled].focus,
  .btn-primary[disabled]:focus,
  .btn-primary[disabled]:hover,
  fieldset[disabled] .btn-primary.focus,
  fieldset[disabled] .btn-primary:focus,
  fieldset[disabled] .btn-primary:hover {
    background-color: #404040;
  }

  .btn-primary:hover,
  .btn-primary:active:focus,
  .btn-primary:focus,
  .btn-primary:active:hover {
    background-color: #278626;
    border-color: #278626;
  }

  .btn-danger {
    background-color: #d9534f;
    border: 1px #d9534f solid;
  }

  .btn-danger:hover,
  .btn-danger:active:hover {
    background-color: #d43f3a;
    border: 1px #d43f3a solid;
  }

  .btn:disabled,
  .btn:disabled:hover,
  .btn:disabled:active:hover {
    background-color: #aaa;
    border-color: #aaa;
  }

  .btn-default {
    background-color: transparent;
    color: #ccc;
    border: 1px solid #d7d7d7;
  }

  .btn-default:focus,
  .btn-default:active,
  .btn-default:hover,
  .btn-default:active:focus,
  .btn-default:hover:focus,
  .btn-default:hover:active {
    background-color: rgba(255, 255, 255, 0.1);
    color: #fff;
    border: 1px solid #fff;
  }

  .btn {
    i.fa {
      padding-right: 0.3em;
    }
  }

  .btn.btn-sm {
    height: 24px;
    padding: 0px 8px !important;
    margin-left: 5px;
  }
`;

interface IButtonProps extends React.ComponentPropsWithoutRef<'button'> {
  bsStyle?: 'primary' | 'danger' | 'default' | 'link' | 'warning';
  bsSize?: 'sm' | 'lg';
  loading?: boolean;
  loadingTimeout?: number;
  disabled?: boolean;
  href?: string;
  target?: '_self' | '_blank' | '_parent' | '_top' | string;
  rel?: string;
}

const Button = React.forwardRef<HTMLButtonElement, IButtonProps>(
  (
    { loading, disabled, bsStyle, bsSize, children, loadingTimeout, ...props },
    ref
  ) => {
    return (
      <Wrapper
        ref={ref as React.RefObject<HTMLDivElement>}
        className='button-wrapper'
      >
        <BsButton
          bsSize={bsSize}
          bsStyle={bsStyle}
          disabled={disabled || loading}
          role='button'
          {...(props as React.ComponentPropsWithoutRef<typeof BsButton>)}
        >
          {children}
        </BsButton>

        <LoadingIndicator
          loading={loading}
          loadingPosition='right'
          timeout={loadingTimeout}
        />
      </Wrapper>
    );
  }
);

Button.propTypes = {
  bsStyle: PropTypes.string as PropTypes.Requireable<IButtonProps['bsStyle']>,
  bsSize: PropTypes.string as PropTypes.Requireable<IButtonProps['bsSize']>,
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.string,
  loadingTimeout: PropTypes.number,
};

Button.defaultProps = {
  bsStyle: 'default',
};

export default Button;
