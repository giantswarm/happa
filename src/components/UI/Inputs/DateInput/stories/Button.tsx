import { Story } from '@storybook/react/types-6-0';
import format from 'date-fns/fp/format';
import parseISO from 'date-fns/fp/parseISO';
import React, { ComponentPropsWithoutRef, useState } from 'react';

import DateInput from '..';

export const Button: Story<ComponentPropsWithoutRef<typeof DateInput>> = (
  args
) => {
  const [value, setValue] = useState('');

  return (
    <DateInput
      {...args}
      value={value}
      onChange={(e) => {
        setValue(e.value as string);
      }}
      buttonProps={{
        label: value
          ? format('d MMM yyyy')(parseISO(value))
          : 'Pick a date and call me Jim',
      }}
    />
  );
};

Button.args = {
  id: 'dob',
};

Button.argTypes = {
  format: { control: { type: 'text' } },
  label: { control: { type: 'text' } },
  error: { control: { type: 'text' } },
  info: { control: { type: 'text' } },
  help: { control: { type: 'text' } },
};
