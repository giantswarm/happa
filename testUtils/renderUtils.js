import { render } from '@testing-library/react';
import App from 'App';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'emotion-theming';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { AppRoutes } from 'shared/constants/routes';
import configureStore from 'stores/configureStore';
import theme from 'styles/theme';

export const initialStorage = {
  user:
    '{"email":"developer@giantswarm.io","auth":{"scheme":"giantswarm","token":"a-valid-token"},"isAdmin":true}',
};

/**
 * This function will render the whole app with a mocked store in the route
 * provided.
 * @param {String} route.
 * @param {HTMLElement} container An HTMLElement.
 */
export function renderRouteWithStore(
  initialRoute = AppRoutes.Home,
  state = {},
  storage = initialStorage
) {
  localStorage.replaceWith(storage);

  const history = createMemoryHistory({
    initialEntries: [initialRoute],
    initialIndex: 0,
  });

  const store = configureStore(state, history);

  const app = render(<App {...{ store, theme, history }} />);

  return app;
}

/**
 * Render component in the Theme context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithTheme(component, props, options) {
  return render(getComponentWithTheme(component, props), options);
}

/**
 * Include component in the Theme context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 */
export function getComponentWithTheme(Component, props) {
  return (
    <ThemeProvider theme={theme}>
      <Component {...props} />
    </ThemeProvider>
  );
}

/**
 * Render component in the Store context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithStore(component, props, state, options) {
  return render(getComponentWithStore(component, props, state), options);
}

/**
 * Include component in the Store context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {Record<string, any>} storage Current LocalStorage value
 * @param {History} history Current Browser history
 */
export function getComponentWithStore(
  Component,
  props = {},
  state = {},
  storage = initialStorage,
  history = createMemoryHistory()
) {
  localStorage.replaceWith(storage);

  const store = configureStore(state, history);

  const app = (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </ThemeProvider>
    </Provider>
  );

  return app;
}
