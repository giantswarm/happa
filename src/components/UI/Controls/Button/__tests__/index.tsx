import { renderWithTheme } from 'test/renderUtils';

import Button from '..';

describe('Button', () => {
  it('renders without crashing', () => {
    const { container } = renderWithTheme(Button, {
      children: 'Hi friends',
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a primary button', () => {
    const { container } = renderWithTheme(Button, {
      children: 'Hi friends',
      primary: true,
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a secondary button', () => {
    const { container } = renderWithTheme(Button, {
      children: 'Hi friends',
      secondary: true,
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a danger button', () => {
    const { container } = renderWithTheme(Button, {
      children: 'Hi friends',
      danger: true,
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a warning button', () => {
    const { container } = renderWithTheme(Button, {
      children: 'Hi friends',
      warning: true,
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a link button', () => {
    const { container } = renderWithTheme(Button, {
      children: 'Hi friends',
      link: true,
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a button with a loading animation', () => {
    const { container } = renderWithTheme(Button, {
      children: 'Hi friends',
      loading: true,
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a disabled button', () => {
    const { container } = renderWithTheme(Button, {
      children: 'Hi friends',
      disabled: true,
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
