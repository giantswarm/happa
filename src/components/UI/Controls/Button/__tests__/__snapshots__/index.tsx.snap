// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders a button with a loading animation 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 giQQcM"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 kuNpSX"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 kshQkK Button__StyledBox-sc-160jtek-0 gfxCxG"
    >
      <button
        class="StyledButtonKind-sc-1vhfpnt-0 eIAcFH"
        disabled=""
        type="button"
      >
        Hi friends
      </button>
      <div
        aria-busy="true"
        aria-hidden="false"
        aria-live="assertive"
        aria-valuetext="Loading…"
        class="LoadingIndicator__ProgressButtonStatusIndicator-sc-yaqsha-0 dYoKvr Button__StyledLoadingIndicator-sc-160jtek-1 RfLMa"
        role="progressbar"
      >
        <img
          alt="loading"
          class="loader right"
          src="[object Object]"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Button renders a danger button 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 giQQcM"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 kuNpSX"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 kshQkK Button__StyledBox-sc-160jtek-0 gfxCxG"
    >
      <button
        class="StyledButtonKind-sc-1vhfpnt-0 dBeDxK"
        type="button"
      >
        Hi friends
      </button>
      <div
        aria-busy="true"
        aria-hidden="true"
        aria-live="assertive"
        aria-valuetext="Loading…"
        class="LoadingIndicator__ProgressButtonStatusIndicator-sc-yaqsha-0 dYoKvr Button__StyledLoadingIndicator-sc-160jtek-1 RfLMa"
        role="progressbar"
      />
    </div>
  </div>
</div>
`;

exports[`Button renders a disabled button 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 giQQcM"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 kuNpSX"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 kshQkK Button__StyledBox-sc-160jtek-0 gfxCxG"
    >
      <button
        class="StyledButtonKind-sc-1vhfpnt-0 eIAcFH"
        disabled=""
        type="button"
      >
        Hi friends
      </button>
      <div
        aria-busy="true"
        aria-hidden="true"
        aria-live="assertive"
        aria-valuetext="Loading…"
        class="LoadingIndicator__ProgressButtonStatusIndicator-sc-yaqsha-0 dYoKvr Button__StyledLoadingIndicator-sc-160jtek-1 RfLMa"
        role="progressbar"
      />
    </div>
  </div>
</div>
`;

exports[`Button renders a link button 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 giQQcM"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 kuNpSX"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 kshQkK Button__StyledBox-sc-160jtek-0 gfxCxG"
    >
      <button
        class="StyledButtonKind-sc-1vhfpnt-0 goECDo"
        type="button"
      >
        Hi friends
      </button>
      <div
        aria-busy="true"
        aria-hidden="true"
        aria-live="assertive"
        aria-valuetext="Loading…"
        class="LoadingIndicator__ProgressButtonStatusIndicator-sc-yaqsha-0 dYoKvr Button__StyledLoadingIndicator-sc-160jtek-1 RfLMa"
        role="progressbar"
      />
    </div>
  </div>
</div>
`;

exports[`Button renders a primary button 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 giQQcM"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 kuNpSX"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 kshQkK Button__StyledBox-sc-160jtek-0 gfxCxG"
    >
      <button
        class="StyledButtonKind-sc-1vhfpnt-0 gXTwvc"
        type="button"
      >
        Hi friends
      </button>
      <div
        aria-busy="true"
        aria-hidden="true"
        aria-live="assertive"
        aria-valuetext="Loading…"
        class="LoadingIndicator__ProgressButtonStatusIndicator-sc-yaqsha-0 dYoKvr Button__StyledLoadingIndicator-sc-160jtek-1 RfLMa"
        role="progressbar"
      />
    </div>
  </div>
</div>
`;

exports[`Button renders a secondary button 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 giQQcM"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 kuNpSX"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 kshQkK Button__StyledBox-sc-160jtek-0 gfxCxG"
    >
      <button
        class="StyledButtonKind-sc-1vhfpnt-0 fBgJoh"
        type="button"
      >
        Hi friends
      </button>
      <div
        aria-busy="true"
        aria-hidden="true"
        aria-live="assertive"
        aria-valuetext="Loading…"
        class="LoadingIndicator__ProgressButtonStatusIndicator-sc-yaqsha-0 dYoKvr Button__StyledLoadingIndicator-sc-160jtek-1 RfLMa"
        role="progressbar"
      />
    </div>
  </div>
</div>
`;

exports[`Button renders a warning button 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 giQQcM"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 kuNpSX"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 kshQkK Button__StyledBox-sc-160jtek-0 gfxCxG"
    >
      <button
        class="StyledButtonKind-sc-1vhfpnt-0 ctvjYB"
        type="button"
      >
        Hi friends
      </button>
      <div
        aria-busy="true"
        aria-hidden="true"
        aria-live="assertive"
        aria-valuetext="Loading…"
        class="LoadingIndicator__ProgressButtonStatusIndicator-sc-yaqsha-0 dYoKvr Button__StyledLoadingIndicator-sc-160jtek-1 RfLMa"
        role="progressbar"
      />
    </div>
  </div>
</div>
`;

exports[`Button renders without crashing 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 giQQcM"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 kuNpSX"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 kshQkK Button__StyledBox-sc-160jtek-0 gfxCxG"
    >
      <button
        class="StyledButtonKind-sc-1vhfpnt-0 fBgJoh"
        type="button"
      >
        Hi friends
      </button>
      <div
        aria-busy="true"
        aria-hidden="true"
        aria-live="assertive"
        aria-valuetext="Loading…"
        class="LoadingIndicator__ProgressButtonStatusIndicator-sc-yaqsha-0 dYoKvr Button__StyledLoadingIndicator-sc-160jtek-1 RfLMa"
        role="progressbar"
      />
    </div>
  </div>
</div>
`;
